### DISCUSSION TITLE
Serializing google.protobuf.Any for visualization in Foxglove Studio

### DISCUSSION THREAD

**Discussion by tervay-bdai** [foxglove/mcap#1115] [General] [answered]
I'm attempting to (in C++) serialize a protobuf utilizing `google.protobuf.Any` as a type, in order to view in Foxglove. For example:

```
message Foo {
  google.protobuf.Any bar = 1;
}

message Baz {
  repeated int32 buzz = 1;
}
```

I would like to insert a `Baz` instance into `Foo::bar` and plot `foo.bar.buzz` in Foxglove. Unfortunately, Foxglove doesn't recognize any schemas for the type:

![image](https://github.com/foxglove/community/assets/132940363/811bc917-bf1a-4dce-bb98-a33b9be42925)

I suspect this is because I am not actually ever telling mcap about the protobuf schemas from the message types packed into the `Any` - I'm only registering `Foo`'s schema, which doesn't know anything about any other types, because, well, `bar` can be anything.

I've taken a read through the [mcap protobuf guide](https://mcap.dev/guides/cpp/protobuf) and the [examples](https://github.com/foxglove/mcap/tree/main/cpp/examples/protobuf), but I'm not totally seeing a way to register multiple proto schemas.

Is this a supported workflow? If so, how would one achieve this?

**Comment by jtbandes** [comment, marked as answer]
Hi,
Currently, we don't have any support for `google.protobuf.Any` and you would need to create a separate schema and channel for each type. One difficulty in supporting `google.protobuf.Any` is that Foxglove Studio needs to know the whole protobuf descriptor (formatted as a FileDescriptorSet) in order to deserialize message data. Simply `type_url: "type.googleapis.com/bdai.mcap.Child"` is not enough information to deserialize messages, unless the app can somehow access the full descriptor for `bdai.mcap.Child` somewhere.

**Reply by tervay-bdai** [reply]
Thank you for the info. I think the way to go about this is - instead of embedding messages within another to log them both, simply log them side-by-side in different channels and make sure the `logTime`/`publishTime` are the same for each pair.
